# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def addTwo(x):
    return x + 2

print(addTwo(4))


# Lambda Function Syntax
# Example of calling a lambda function without a variable
print((lambda x: x + 2)(4))

#Single line execution- must be a one-liner.  Way more restrictive than a general function definition

6
6


#### Saving to a Variable

In [3]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [5]:
# Multiple Inputs with no variable
# format ==> lambda parameters: return-line

print((lambda x,y,z: x * y * z)(3,5,8)) # can run w/o parameters to show error

# Multiple Inputs with a variable attached
x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [1]:
# we're setting this func up to accept a function!
def multiply(f,num):
    """
        f expects a lambda function
        num expects a number
    """
    return f(num)


multiply(lambda x: x * x, 4) # lambda 4: 4 * 4

#we can change this for each execution!!!
multiply(lambda x : x + 10, 34)

44

#### Returning a Lambda from a Function

In [6]:


# Regular defined function
def multiply_test(num):
    return num * 4

# Function within a function
def returnFunc():
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))


# BIG EXAMPLE of closure!
def test():
    names = [] #discuss scope, access here
    s = lambda name: names.append(name)
#     first iteration with x, need arguments with this one
    t = lambda x: print(names)
#     second iteration without x
    return s, t

# destructuring here as we have to values:
addNames, printInfo = test()
addNames('Brendan')
addNames('Brandt')
addNames('Rachel')
printInfo('234') # if arguments needed here we can put in whatever
printInfo() #no arguments example

#closure, try to print private/aka protected variable for example
    

# final example:
# Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x,a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb)
print(type(r_lamb))
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x000001E2E9BDFB80>
8
['Brendan', 'Brandt', 'Rachel']
<function returnLam.<locals>.<lambda> at 0x000001E2E9C350D0>
<class 'function'>
19


#### If Statements within Lambdas

In [22]:
# Lambda x : True if (condition) else False

f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))

print(f_condition(12))

print(f_condition(10))

#In case someone asks about elif. . . nesting in the else can work
# lambda x: x*10 if x<2 else (x**2 if x<4 else x+10)

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [17]:
# map(func,iterable(list,dict,tuple,set))
# Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
# let's start with a list and square all the nums
numbers = [4,11,20,3,15,20]
other_nums = [2, 4, 6, 8, 10]
# defining a func
def squared(num):
    return num ** 2
# loops through list and applies func
for n in numbers:
    n = squared(n)
print(numbers)
    
# with a list comp
next_nums = [n**2 for n in numbers]
print(next_nums)


# Finally with the map function:
# first with the pre-defined func
map(squared, other_nums) #shows the map object
for m in map(squared, other_nums): #another bit to show that we could and what the map object is
    print(m)
list(map(squared, other_nums)) #our normal use case


[4, 11, 20, 3, 15, 20]
[16, 121, 400, 9, 225, 400]
4
16
36
64
100


[4, 16, 36, 64, 100]

#### Using Lambda's with Map

In [11]:
list(map(lambda x: x**2, other_nums))

#nice compact version of all the other things we've been showing with this one!


[4, 16, 36, 64, 100]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [1]:
# 

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [18]:
# syntax-->   filter(func, iterable)

som_nums = [1, 2, 3, 4, 5]
def filtOdds(num):
    if num % 2 == 0: #after example, let's turn this to return num%2==0 to show it
        return True
    return False
print(list(filter(filtOdds, other_nums)))

# Another example with strings:

names = ['Bob','Andy','Max','Evan','Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
    # re-factor to one-line return
list(filter(a_names, names))




[2, 4, 6, 8, 10]


['Andy', 'Angelica']

#### Using Lambda's with Filter()

In [19]:
saved_filter = list(filter(lambda name: name[0].lower() == 'a', names))
print(saved_filter)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [2]:
from statistics import mean



## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [67]:
from functools import reduce

# reduce(function,iterable)
# The whole point of reduce is to aggregate all items in a list to one value (like sum() min() etc.)
# This function should take in two parameters and return 1 like below

list_1 = [2,4,6,8,10]

def addNums(num1,num2):
    return num1 + num2

result_add = reduce(addNums,list_1)

print(result_add)

# Subtract a list of numbers
def subtractNums(num1,num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)


30
-26


#### Using Lambda's with Reduce()

In [68]:
result_lamb = reduce(lambda x,y: x + y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

## .Sort() w/ lambda!


In [ ]:
# list.sort(reverse=True|False, key=myFunc)

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [29]:
# First one is an infinite loop:
# def addNums(num):
#     return num + addNums(num+1)
# addNums(1)
                    #Because: 1: 1+addNums(2)
                    #         2: 2+addNums(3)
                    #         3: 3+addNums(4). . .  etc forever!

# SOOoooo we need to add a base case!
# def addNums(num):
#     if num >= 10:
#         print('addNums(10) = 10')
#         return num
#     else:
#         print(f"addNums({num}) = {num} + addNums({num+1})")
#         return num + addNums(num+1)
                    
# addNums(1)

# SSSOOOOOOO:  First we have to perform these operations all the way down until we hit our base case and THEN
# we can say- "hey, we finally have a solution for addNums(10)!  It's 10! let's plug that in backwards down the line"
#     addNums(1) = 1 + addNums(2)
#     addNums(2) = 2 + addNums(3)
#     addNums(3) = 3 + addNums(4)
#     addNums(4) = 4 + addNums(5)
#     addNums(5) = 5 + addNums(6)
#     addNums(6) = 6 + addNums(7)
#     addNums(7) = 7 + addNums(8)
#     addNums(8) = 8 + addNums(9)  -->addnums(9) is now 9! so we add 8 to that!
#     addNums(9) = 9 + addNums(10) -->addNums(10) is now 10! so we add 9 to that!
#     addNums(10) = 10

# def addNums(num):
#     # Set base case for recusive function!!!!
#     if num <= 1:
#         print("addNums(1) = 1")
#         return num
        # recursive case!!!!
#     else:
#         print(f"addNums({num}) = {num} + addNums({num - 1})")
#         return num + addNums(num - 1)
# addNums(5)

# Final notes: similar idea to while loop in incrementing counts so you don't have infinite loops

addNums(1) = 1 + addNums(2)
addNums(2) = 2 + addNums(3)
addNums(3) = 3 + addNums(4)
addNums(4) = 4 + addNums(5)
addNums(5) = 5 + addNums(6)
addNums(6) = 6 + addNums(7)
addNums(7) = 7 + addNums(8)
addNums(8) = 8 + addNums(9)
addNums(9) = 9 + addNums(10)
addNums(10) = 10


55

#### Writing a Factorial Function

In [3]:
# 5! = 5 * 4 * 3 * 2 * 1
# just to show the while loop version:
def factorial2(num):
    count = 1
    while num >=1:
        print(count, num)
        count = num * count
        num -= 1
    return count
    
print(factorial2(5))
    
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
    
factorial(5)


1 5
5 4
20 3
60 2
120 1
120


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [40]:
range(1,10)
print(range(1,10))
print(type(range(1,10)))
# help(range)
print(range(1,10)[4])

# Generators aren't storing the data, it's storing the formula to generate it on the fly.  Kind of like a for loop.
# in this exp:  num is never stored it's just spit out and overwritten on the next iteration-->
# for num in range(10000):
#     num. . . 

# now let's make one!

range(1, 10)
<class 'range'>
5


In [43]:
def my_range(start, stop, step ):
    while start < stop:
        yield start # yield keyword denotes a generator and generally means keep returning.
        start += step
# my_range(2, 22, 2)# show the generator object
# my_range(2, 22, 2)[5] #not indexed!!!  Range is. . . 

for i in my_range(5, 50, 5):
    print(i)



5
10
15
20
25
30
35
40
45


#### Infinite Generator

In [ ]:
# bad, never create infinite loops
# create a stop!

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [52]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
output = ['Argentina', 'San Diego', 'Boston', 'New York'] #yet another example 'New Cork'
print(places)

while " " in places:
    places.remove(" ")
    places.remove("")
   
   
print(places)

print(sorted(output, key=lambda p: p.split()[-1]))

[' ', 'Argentina', ' ', 'San Diego', '', '  ', '', 'Boston', 'New York']
['Argentina', 'San Diego', '  ', 'Boston', 'New York']
['Argentina', 'Boston', 'New Cork', 'San Diego']


In [1]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

print ("Old List : " + str(places))
while("" in places) :
    places.remove("")
while(" " in places) :
    places.remove(" ")
while("" in places) :
    places.remove("")
while("  " in places) :
    places.remove("  ")
print ("New List : " + str(places))

Old List : [' ', 'Argentina', ' ', 'San Diego', '', '  ', '', 'Boston', 'New York']
New List : ['Argentina', 'San Diego', 'Boston', 'New York']


In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

places_new = list(filter(lambda x: len(x)> 4, places))

print(places_new)

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [2]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda last: last.split()[-1].lower())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


In [3]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

sorted(author, key=lambda x: x.split(" ")[-1].lower())

['Victor aNisimov',
 'Gary A.J. Bernstein',
 'Joel Carter',
 'Andrew P. Garfield',
 'David hassELHOFF']

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [8]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
my_places = list(map(lambda c: (c[0], (9/5)* c[1]+32), places))
print(my_places)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


In [4]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

import numpy as np
inp = [32,12,44,29]

cel = np.array(inp)
 
print(f"Celsius {cel}")

feh = (9 * cel / 5 + 32)
 
print(f"Fahrenheit {feh}")
print(places)

Celsius [32 12 44 29]
Fahrenheit [ 89.6  53.6 111.2  84.2]
[('Nashua', 32), ('Boston', 12), ('Los Angelos', 44), ('Miami', 29)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [1]:
def fibNum(n):
    step = 1
    if n <= 1:
        return 1
    else:
        return fibNum(n-2) + fibNum(n-1)
    
def fibSeq(num):
    count = 0
    for x in range(num +2)[1:]:
        print(f"Iteration {count}: {fibNum(x)}")
        count +=1
        
print(fibSeq(5))
print(fibNum(8))

Iteration 0: 1
Iteration 1: 2
Iteration 2: 3
Iteration 3: 5
Iteration 4: 8
Iteration 5: 13
None
34


In [1]:
def recur_fib(num):
    # Set base case for recusive function
    if num <= 1:
        return num
    else:
        return recur_fib(num-1) + recur_fib(num-2)

        print(f"Iteration {num}: {recur_fib(num-1) + recur_fib (num-2)}")
recur_fib(5)

89

In [11]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]


author.sort(key = lambda l_name: l_name.split(" ")[-1].lower())
print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


In [9]:

def fib(num):
    if num > 1:
        return fib(num-1) + fib(num-2)
    return num
for i in range(11):
    print(f"Iteration {i}: {fib(i)}")

Iteration 0: 0
Iteration 1: 1
Iteration 2: 1
Iteration 3: 2
Iteration 4: 3
Iteration 5: 5
Iteration 6: 8
Iteration 7: 13
Iteration 8: 21
Iteration 9: 34
Iteration 10: 55


In [ ]:
new_places = list(filter(lambda place: place.strip(), places))

In [5]:
def fib(num):
    if num <= 1:
        return num
    else:
        return(fib(num-1) + fib(num-2))

nums = 20

for n in range(nums):
    print(fib(n))

0
1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181


In [3]:
places = ['', '   ', 'bob']
def ns():
    return p and not

bob


In [ ]:
class Cart:
    def __init__(self):
        self.items = []

    def add_items(self):
        name = input("What would you like to add? ")
        number = int(input("How many would you like to add? "))
        price = input("What is the price? ")

        new_item = Item(name, number, price)
        print(new_item, type(new_item))
        print(new_item.__dict__)
        self.items.append(new_item)

    def remove_item(self):
        remove_item = input('Which item would you like to remove? ')
        remove_number= input('How many? ')
        for i in range(len(self.items)):
            if self.items[i].name.lower() == remove_item.lower():
                if remove_number.isdigit():
                    remove_number = int(remove_number)
                    self.items[i].number -= remove_number
                    print(f'{remove_number} {remove_item} were removed.')
                    if self.items[i].number < 1:
                        self.items.pop(i)
                        print(f'{remove_item} was removed')
                        break
                    else:
                        break
                print('Please enter number in digits')
            else:
                print(f'{remove_item} not found')

    
    def view_items(self):
        for item in self.items:
            print(f'{item.number} {item.name} at ${item.price}')
    def run(self):
        while True:
           while True:
            user_choice = input("What would you like to do? (add/remove/view/quit)").lower()
            
            if user_choice == 'add':
                self.add_items()
            elif user_choice == 'remove':
                self.remove_item()
            elif user_choice == 'view':
                self.view_items()
            elif user_choice == 'quit':
                self.view_items()
                return
            else:
                print('input unsupported, please try again') 



class Item:
    def __init__(self, name, number, price):
        self.name = name
        self.number = number
        self.price = price

shop = Cart()

shop.run()

What would you like to do? (add/remove/view/quit)add
What would you like to add? tomato
How many would you like to add? 23
What is the price? 56
<__main__.Item object at 0x0000026890563400> <class '__main__.Item'>
{'name': 'tomato', 'number': 23, 'price': '56'}
